**Note: Run servers on local instead of Colab**

Here are some comments for a tutorial on Model serialization with Pickle/Joblib:

1. **Introduction to Model Serialization**:
   - Model serialization is the process of converting a machine learning model into a format that can be easily saved and loaded. This is crucial for deploying models in production environments or sharing them with others
.

2. **Why Use Pickle or Joblib?**:
   - Pickle and Joblib are two popular libraries in Python for serializing machine learning models. Pickle is a built-in Python module that can serialize and deserialize any Python object, including custom classes and objects
.
   - Joblib is more efficient than Pickle when working with large numpy arrays, making it a better choice for serializing large machine learning models
.

3. **Security Considerations**:
   - Both Pickle and Joblib use the pickle protocol under the hood, which can execute arbitrary code during unpickling. Therefore, it is important to only load models from trusted sources to avoid security risks
.

4. **Saving a Model with Pickle**:
   - To save a model using Pickle, you can use the `pickle.dump` function. This function serializes the model and writes it to a file. For example:
     ```python
     import pickle
     with open('model.pkl', 'wb') as file:
         pickle.dump(model, file)
     ```
   - This code opens a file in write-binary mode and uses Pickle to serialize the model into this file
.

5. **Loading a Model with Pickle**:
   - To load a model using Pickle, you can use the `pickle.load` function. This function reads the serialized model from a file and deserializes it. For example:
     ```python
     with open('model.pkl', 'rb') as file:
         model = pickle.load(file)
     ```
   - This code opens the file in read-binary mode and uses Pickle to deserialize the model from this file
.

6. **Saving a Model with Joblib**:
   - To save a model using Joblib, you can use the `joblib.dump` function. This function serializes the model and writes it to a file. For example:
     ```python
     from joblib import dump, load
     dump(model, 'model.joblib')
     ```
   - This code uses Joblib to serialize the model into a file named 'model.joblib'
.

7. **Loading a Model with Joblib**:
   - To load a model using Joblib, you can use the `joblib.load` function. This function reads the serialized model from a file and deserializes it. For example:
     ```python
     model = load('model.joblib')
     ```
   - This code uses Joblib to deserialize the model from the file named 'model.joblib'
.

8. **Performance Comparison**:
   - Joblib is generally faster than Pickle when dealing with large numpy arrays due to its optimizations for such data structures. However, Pickle may be faster for smaller models or models that do not contain large numpy arrays
.

9. **Use Cases**:
   - Use Pickle when you need to serialize small models or models that do not contain large numpy arrays. Use Joblib when you need to serialize large models or models that contain large numpy arrays
.

10. **Conclusion**:
    - Both Pickle and Joblib are powerful tools for model serialization. The choice between them depends on the size of your model and the data structures it contains. Always ensure that you load models from trusted sources to avoid security risks
.

In [1]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
X, y = load_breast_cancer(return_X_y=True)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [2]:
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier
pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('clf', RandomForestClassifier(n_estimators=100, random_state=42))
])
pipeline.fit(X_train, y_train)


Pipeline(steps=[('scaler', StandardScaler()),
                ('clf', RandomForestClassifier(random_state=42))])

In [3]:
import pickle
version = '1.0.0'
filename_pkl = f'model_v{version}.pkl'
with open(filename_pkl, 'wb') as f:
    pickle.dump(pipeline, f)


In [4]:
import joblib
filename_joblib = f'model_v{version}.joblib'
joblib.dump(pipeline, filename_joblib)


['model_v1.0.0.joblib']

In [5]:
with open(filename_pkl, 'rb') as f:
    model_pickle = pickle.load(f)
preds_pickle = model_pickle.predict(X_test)


In [6]:
with open(filename_pkl, 'rb') as f:
    model_pickle = pickle.load(f)
preds_pickle = model_pickle.predict(X_test)


In [7]:
model_joblib = joblib.load(filename_joblib)
preds_joblib = model_joblib.predict(X_test)


In [8]:
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test, preds_pickle))
print(accuracy_score(y_test, preds_joblib))


0.9649122807017544
0.9649122807017544


In [9]:
import os
versions = ['1.0.0', '1.1.0', '2.0.0']
for v in versions:
    fname = f'model_v{v}.joblib'
    joblib.dump(pipeline, fname)
os.listdir('.')


['.config',
 'model_v1.0.0.pkl',
 'model_v2.0.0.joblib',
 'model_v1.0.0.joblib',
 'model_v1.1.0.joblib',
 'sample_data']

In [10]:
import json
metadata = {
    'model_name': 'RandomForestPipeline',
    'versions': versions,
    'date_saved': {v: f'model_v{v}.joblib' for v in versions}
}
with open('model_metadata.json', 'w') as f:
    json.dump(metadata, f)


In [11]:
with open('model_metadata.json', 'r') as f:
    meta = json.load(f)
print(meta)


{'model_name': 'RandomForestPipeline', 'versions': ['1.0.0', '1.1.0', '2.0.0'], 'date_saved': {'1.0.0': 'model_v1.0.0.joblib', '1.1.0': 'model_v1.1.0.joblib', '2.0.0': 'model_v2.0.0.joblib'}}


Excercise

### Overview  
This exercise guides you through training a scikit-learn pipeline, saving it with both Pickle and Joblib, managing model versions and metadata, and exposing the model via a simple Flask API  ([9. Model persistence — scikit-learn 1.6.1 documentation](https://scikit-learn.org/stable/model_persistence.html?utm_source=chatgpt.com)).  

### Tasks  
1. **Data Preparation:** Load the Breast Cancer Wisconsin dataset (`sklearn.datasets.load_breast_cancer`), split into train/test (80/20), and standardize features  ([Save and Load Machine Learning Models in Python with scikit-learn](https://machinelearningmastery.com/save-load-machine-learning-models-python-scikit-learn/?utm_source=chatgpt.com)).  
2. **Pipeline Construction:** Build an `sklearn.pipeline.Pipeline` with `StandardScaler` and `RandomForestClassifier(n_estimators=100, random_state=42)` and fit it on the training data  ([How to properly pickle sklearn pipeline when using custom ...](https://stackoverflow.com/questions/57888291/how-to-properly-pickle-sklearn-pipeline-when-using-custom-transformer?utm_source=chatgpt.com)).  
3. **Serialization:**  
   - Save the fitted pipeline using Pickle protocol 5 to `model_v1.pkl`.  
   - Save the same pipeline using Joblib to `model_v1.joblib`  ([Save Machine Learning Model Using Pickle and Joblib](https://www.analyticsvidhya.com/blog/2021/08/quick-hacks-to-save-machine-learning-model-using-pickle-and-joblib/?utm_source=chatgpt.com)).  
4. **Version Control:** Simulate versioning by retraining the pipeline with a different random seed, then saving as `model_v2.pkl`/`.joblib`; repeat for a third version. Create a `model_metadata.json` cataloging versions, file names, and timestamps  ([Machine Learning Model Serialization - Christopher Flynn](https://flynn.gg/blog/machine-learning-model-serialization/?utm_source=chatgpt.com)).  
5. **Deserialization & Validation:**  
   - Load each Pickle and Joblib file, run `.predict()` on the test set, and report accuracy.  
   - Measure and compare file sizes and load times between Pickle and Joblib  ([Pickle is over 10 times faster than joblib for save and load scikit ...](https://www.reddit.com/r/Python/comments/ypj13g/pickle_is_over_10_times_faster_than_joblib_for/?utm_source=chatgpt.com)).  
6. **Model Serving (Bonus):**  
   - Implement a minimal Flask app with two endpoints:  
     - `GET /models` returns available version metadata.  
     - `POST /predict` accepts JSON feature vectors and returns predicted classes using the latest model.  
   - Demonstrate sending requests via `requests` in Colab  ([Model Serialization using pickle and joblib - Kaggle](https://www.kaggle.com/code/tasnimniger/model-serialization-using-pickle-and-joblib?utm_source=chatgpt.com)).  

### Deliverables  
- A Colab notebook containing all code cells above.  
- The saved model files (`.pkl`, `.joblib`) and `model_metadata.json`.  
- A brief report (markdown cell) comparing Pickle vs Joblib in speed and size.  
- (Optional) The Flask app code cell demonstrating model serving.

**Data Preparation**

In [12]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split

X, y = load_breast_cancer(return_X_y=True)
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

print("Train shape:", X_train.shape)
print("Test shape:", X_test.shape)

Train shape: (455, 30)
Test shape: (114, 30)


**Pipeline Built & Fit**

In [13]:
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier

pipeline_v1 = Pipeline([
    ('scaler', StandardScaler()),
    ('clf', RandomForestClassifier(n_estimators=100, random_state=42))
])

pipeline_v1.fit(X_train, y_train)
print("Pipeline v1 trained successfully")

Pipeline v1 trained successfully


**Save Model With Pikle & Joblib **

In [14]:
import pickle
import joblib

# Save with Pickle (protocol 5)
filename_pkl_v1 = 'model_v1.pkl'
with open(filename_pkl_v1, 'wb') as f:
    pickle.dump(pipeline_v1, f, protocol=5)

# Save with Joblib
filename_joblib_v1 = 'model_v1.joblib'
joblib.dump(pipeline_v1, filename_joblib_v1)

print("Saved:", filename_pkl_v1, "and", filename_joblib_v1)

Saved: model_v1.pkl and model_v1.joblib


**Version Control**

In [16]:
import json
from datetime import datetime

# ---- Version 2 (different seed) ----
pipeline_v2 = Pipeline([
    ('scaler', StandardScaler()),
    ('clf', RandomForestClassifier(n_estimators=100, random_state=7))
])
pipeline_v2.fit(X_train, y_train)

with open('model_v2.pkl', 'wb') as f:
    pickle.dump(pipeline_v2, f, protocol=5)
joblib.dump(pipeline_v2, 'model_v2.joblib')

# ---- Version 3 (another seed) ----
pipeline_v3 = Pipeline([
    ('scaler', StandardScaler()),
    ('clf', RandomForestClassifier(n_estimators=100, random_state=99))
])
pipeline_v3.fit(X_train, y_train)

with open('model_v3.pkl', 'wb') as f:
    pickle.dump(pipeline_v3, f, protocol=5)
joblib.dump(pipeline_v3, 'model_v3.joblib')

print("v2 and v3 trained and saved")

# ---- Metadata catalog ----
metadata = {
    'model_name': 'RandomForestPipeline',
    'versions': {
        'v1': {
            'random_state': 42,
            'pkl_file': 'model_v1.pkl',
            'joblib_file': 'model_v1.joblib',
            'date_saved': datetime.now().isoformat()
        },
        'v2': {
            'random_state': 7,
            'pkl_file': 'model_v2.pkl',
            'joblib_file': 'model_v2.joblib',
            'date_saved': datetime.now().isoformat()
        },
        'v3': {
            'random_state': 99,
            'pkl_file': 'model_v3.pkl',
            'joblib_file': 'model_v3.joblib',
            'date_saved': datetime.now().isoformat()
        }
    }
}

with open('model_metadata.json', 'w') as f:
    json.dump(metadata, f, indent=4)

print("model_metadata.json created")

v2 and v3 trained and saved
model_metadata.json created


**Load Predict & Compare**

In [17]:
import time
import os
from sklearn.metrics import accuracy_score

versions = ['v1', 'v2', 'v3']
results = []

for v in versions:
    pkl_file = f'model_{v}.pkl'
    joblib_file = f'model_{v}.joblib'

    # --- Pickle: load time ---
    start = time.time()
    with open(pkl_file, 'rb') as f:
        model_pkl = pickle.load(f)
    pkl_load_time = time.time() - start
    preds_pkl = model_pkl.predict(X_test)
    acc_pkl = accuracy_score(y_test, preds_pkl)
    pkl_size = os.path.getsize(pkl_file) / 1024  # KB

    # --- Joblib: load time ---
    start = time.time()
    model_joblib = joblib.load(joblib_file)
    joblib_load_time = time.time() - start
    preds_joblib = model_joblib.predict(X_test)
    acc_joblib = accuracy_score(y_test, preds_joblib)
    joblib_size = os.path.getsize(joblib_file) / 1024  # KB

    results.append({
        'version': v,
        'pkl_accuracy': acc_pkl,
        'pkl_size_KB': round(pkl_size, 2),
        'pkl_load_time_s': round(pkl_load_time, 5),
        'joblib_accuracy': acc_joblib,
        'joblib_size_KB': round(joblib_size, 2),
        'joblib_load_time_s': round(joblib_load_time, 5),
    })

import pandas as pd
df_results = pd.DataFrame(results)
print(df_results)

  version  pkl_accuracy  pkl_size_KB  pkl_load_time_s  joblib_accuracy  \
0      v1      0.964912       308.29          0.00927         0.964912   
1      v2      0.964912       316.57          0.00294         0.964912   
2      v3      0.964912       316.88          0.00294         0.964912   

   joblib_size_KB  joblib_load_time_s  
0          318.81             0.02092  
1          327.10             0.02350  
2          327.41             0.03207  


**Flask**

In [18]:
from flask import Flask, request, jsonify
import threading
import numpy as np

app = Flask(__name__)

# Latest model use karenge predictions ke liye (v3)
latest_model = joblib.load('model_v3.joblib')

@app.route('/models', methods=['GET'])
def get_models():
    with open('model_metadata.json', 'r') as f:
        meta = json.load(f)
    return jsonify(meta)

@app.route('/predict', methods=['POST'])
def predict():
    data = request.get_json()
    features = np.array(data['features']).reshape(1, -1)
    prediction = latest_model.predict(features)
    return jsonify({'prediction': int(prediction[0])})

def run_app():
    app.run(port=5000, use_reloader=False)

thread = threading.Thread(target=run_app)
thread.start()

import time
time.sleep(2)  # server ko start hone ka time do
print("Flask server started on port 5000")

 * Serving Flask app '__main__'
 * Debug mode: off


INFO:werkzeug:WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on http://127.0.0.1:5000
INFO:werkzeug:Press CTRL+C to quit


Flask server started on port 5000


In [19]:
# request set
import requests

# GET /models test
resp1 = requests.get('http://127.0.0.1:5000/models')
print("GET /models ->", resp1.json())

# POST /predict test — X_test ka pehla sample bhejte hain
sample_features = X_test[0].tolist()
resp2 = requests.post(
    'http://127.0.0.1:5000/predict',
    json={'features': sample_features}
)
print("POST /predict ->", resp2.json())
print("Actual label ->", y_test[0])

INFO:werkzeug:127.0.0.1 - - [24/Jul/2026 19:18:31] "GET /models HTTP/1.1" 200 -
INFO:werkzeug:127.0.0.1 - - [24/Jul/2026 19:18:31] "POST /predict HTTP/1.1" 200 -


GET /models -> {'model_name': 'RandomForestPipeline', 'versions': {'v1': {'date_saved': '2026-07-24T19:14:06.779821', 'joblib_file': 'model_v1.joblib', 'pkl_file': 'model_v1.pkl', 'random_state': 42}, 'v2': {'date_saved': '2026-07-24T19:14:06.779840', 'joblib_file': 'model_v2.joblib', 'pkl_file': 'model_v2.pkl', 'random_state': 7}, 'v3': {'date_saved': '2026-07-24T19:14:06.779844', 'joblib_file': 'model_v3.joblib', 'pkl_file': 'model_v3.pkl', 'random_state': 99}}}
POST /predict -> {'prediction': 1}
Actual label -> 1


## Pickle vs Joblib — Comparison Report

**Accuracy:** Dono methods (Pickle aur Joblib) ne har version (v1, v2, v3) pe bilkul same
accuracy diya (96.49%), kyunke serialization sirf model ko save/load karta hai —
predictions pe koi asar nahi padta.

**File Size:** Joblib files thodi bari thi Pickle se (318–327 KB vs 308–317 KB),
kyunke Joblib numpy arrays ko alag tarike se compress/store karta hai jo chhote
models ke liye extra overhead add kar sakta hai.

**Load Time:** Is case me Pickle Joblib se faster tha (0.003–0.009s vs 0.02–0.03s).
Ye is exercise ke dataset (Breast Cancer — chhota dataset) ke liye ho raha hai;
Joblib generally bade numpy-heavy models (jaise large deep learning weights) pe
zyada efficient hota hai, lekin chhote sklearn models pe Pickle ka overhead kam hota hai.

**Conclusion:** Chhote models (jaise is exercise ka RandomForest) ke liye Pickle
zyada fast hai. Bade, numpy-array-heavy models ke liye Joblib better choice hoti hai.
Dono cases me sirf trusted sources se hi models load karne chahiye (security risk
due to arbitrary code execution during unpickling).